common/error_test.cpp

```diff
@@ -29,94 +31,199 @@ auto IndirectError() -> Error { return Error("test"); }
-    CARBON_RETURN_IF_ERROR(ErrorOr<Success>(Success()));
-    CARBON_RETURN_IF_ERROR(ErrorOr<Success>(Error("error")));
+TYPED_TEST(ErrorOrTest, ReturnIfErrorHasError) {
+  using TestErrorOr = typename TestFixture::template TestErrorOr<Success>;
+  auto result = [this]() -> TestErrorOr {
+    CARBON_RETURN_IF_ERROR(TestErrorOr(Success()));
+    CARBON_RETURN_IF_ERROR(TestErrorOr(this->MakeError()));
     return Success();
   }();
-  EXPECT_THAT(result, IsError("error"));
+  EXPECT_THAT(result, IsError(this->ErrorStr()));
 }
 
-TEST(ErrorTest, AssignOrReturnNoError) {
-  auto result = []() -> ErrorOr<int> {
-    CARBON_ASSIGN_OR_RETURN(int a, ErrorOr<int>(1));
-    CARBON_ASSIGN_OR_RETURN(const int b, ErrorOr<int>(2));
+TYPED_TEST(ErrorOrTest, AssignOrReturnNoError) {
+  using TestErrorOr = typename TestFixture::template TestErrorOr<int>;
+  auto result = []() -> TestErrorOr {
+    CARBON_ASSIGN_OR_RETURN(int a, TestErrorOr(1));
+    CARBON_ASSIGN_OR_RETURN(const int b, TestErrorOr(2));
     int c = 0;
-    CARBON_ASSIGN_OR_RETURN(c, ErrorOr<int>(3));
+    CARBON_ASSIGN_OR_RETURN(c, TestErrorOr(3));
     return a + b + c;
   }();
   EXPECT_THAT(result, IsSuccess(Eq(6)));
 }
 
-TEST(ErrorTest, AssignOrReturnHasDirectError) {
-  auto result = []() -> ErrorOr<int> {
-    CARBON_RETURN_IF_ERROR(ErrorOr<int>(Error("error")));
+TYPED_TEST(ErrorOrTest, AssignOrReturnHasDirectError) {
+  using TestErrorOr = typename TestFixture::template TestErrorOr<int>;
+  auto result = [this]() -> TestErrorOr {
+    CARBON_RETURN_IF_ERROR(TestErrorOr(this->MakeError()));
     return 0;
   }();
-  EXPECT_THAT(result, IsError("error"));
+  EXPECT_THAT(result, IsError(this->ErrorStr()));
 }
 
-TEST(ErrorTest, AssignOrReturnHasErrorInExpected) {
-  auto result = []() -> ErrorOr<int> {
-    CARBON_ASSIGN_OR_RETURN(int a, ErrorOr<int>(Error("error")));
+TYPED_TEST(ErrorOrTest, AssignOrReturnHasErrorInExpected) {
+  using TestErrorOr = typename TestFixture::template TestErrorOr<int>;
+  auto result = [this]() -> TestErrorOr {
+    CARBON_ASSIGN_OR_RETURN(int a, TestErrorOr(this->MakeError()));
     return a;
   }();
-  EXPECT_THAT(result, IsError("error"));
+  EXPECT_THAT(result, IsError(this->ErrorStr()));
 }
 
-TEST(ErrorTest, ErrorBuilderOperatorImplicitCast) {
-  ErrorOr<int> result = ErrorBuilder() << "msg";
-  EXPECT_THAT(result, IsError("msg"));
+class AnotherCustomError : public ErrorBase<AnotherCustomError> {
+ public:
+  auto Print(llvm::raw_ostream& os) const -> void {
+    os << "Another custom test error!";
+  }
+
+  explicit operator CustomError() { return CustomError(); }
+};
+
+TYPED_TEST(ErrorOrTest, AssignOrReturnNoErrorAcrossErrorTypes) {
+  using TestErrorOr = typename TestFixture::template TestErrorOr<int>;
+  auto result = []() -> ErrorOr<int> {
+    CARBON_ASSIGN_OR_RETURN(int a, TestErrorOr(1));
+    CARBON_ASSIGN_OR_RETURN(const int b, []() -> TestErrorOr {
+      CARBON_ASSIGN_OR_RETURN(int inner, (ErrorOr<int, AnotherCustomError>)(2));
```

Comment:
If you're sticking with casting over the `ToError` comment, this should probably follow [casting style](https://google.github.io/styleguide/cppguide.html#Casting), also below

Reply:
(see other comment, this was always intended to be constructor calls... anyways...)